PreservedObjectHandler: refactored results and logging into separate class

ndushay opened the pull request:
This shortens PreservedObjectHandler by about 100 lines (from about 400 to about 300).

Closes #327

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -56,115 +28,110 @@ def initialize(druid, incoming_version, incoming_size, endpoint)
 
-    log_results(results)
-    results
+    handler_results.log_my_results
```
`log_my_results` is a hokey name;  i wanted to distinguish it from the class `log_results` method ... suggestions welcome.

jmartin-sul replied:
i think it'd be reasonable to collapse the two into an instance method called `log_results`, something like:
```ruby
def log_results
  result_array.each do |r|
    severity = logger_severity_level(r.keys.first)
    msg = r.values.first
    Rails.logger.log(severity, msg)
  end
end
```

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -56,115 +28,110 @@ def initialize(druid, incoming_version, incoming_size, endpoint)
+      )
     end
-    errors
+    moab_errors
```
I refactored moab_validation_errors method in addition to the logging changes -- I think it's clearer now.  Happy to entertain dissent.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -185,81 +152,69 @@ def create_db_objects(status, validated=false)
-    else
-      results.concat(transaction_results)
-    end
+    handler_results.add_result(PreservedObjectHandlerResults::CREATED_NEW_OBJECT) if transaction_ok
```
here, having the transaction wrapper method return a boolean works nicely, IMO.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -185,81 +152,69 @@ def create_db_objects(status, validated=false)
-    else
-      full_results.concat(transaction_results)
-    end
+    handler_results.remove_db_updated_result unless transaction_ok
```
here, having the transaction wrapper method return a boolean, and then doing this "remove results" is a little weird ... BUT I think it makes things clearer to the point that I felt John's block comments were no longer needed.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -185,81 +152,69 @@ def create_db_objects(status, validated=false)
+      confirm_version_on_db_object(pc_db_object, :version)
     end
-    results.concat(confirm_results)
+    handler_results.remove_db_updated_result unless transaction_ok
```
adding some actual logic here dependent on transaction results ...

jmartin-sul replied:
slightly surprised/worried this didn't require any tests to be fixed.

ndushay replied:
confirm_version was written early on;  i think our mental models grew over time blah blah.

ndushay commented on spec/services/preserved_object_handler_create_spec.rb:
```diff
@@ -40,13 +40,14 @@
         expect(Rails.logger).to receive(:log).with(Logger::ERROR, exp_msg)
-        po_handler.create
+        new_po_handler = described_class.new(druid, incoming_version, incoming_size, ep)
+        new_po_handler.create
```
this is the one test I had to change -- it's abnormal to re-use a pohandler object so I changed this test so it no longer did.

jmartin-sul replied:
as a stretch-goal/patch-party type thing, maybe we could be nice to the caller and set a flag internally that'd let us raise an error when an instance has been used and a new one has to be instantiated?

ndushay replied:
sure, could be a patch party ticket.  Make it so.  Also make a case for it in the ticket??

jmartin-sul replied:
see #388

jmartin-sul commented on app/services/preserved_object_handler.rb:
```diff
@@ -56,115 +28,110 @@ def initialize(druid, incoming_version, incoming_size, endpoint)
-      log_results(results)
+      PreservedObjectHandlerResults.log_results(
+        [ handler_results.result_hash(PreservedObjectHandlerResults::INVALID_MOAB, moab_error_msgs) ]
+      )
```
maybe:
```ruby
handler_results.add_result(PreservedObjectHandlerResults::INVALID_MOAB, moab_error_msgs)
```

ndushay replied:
probably a typo - will fix.

ndushay replied:
actually - this is where the separate logging was, and you were right about this being part of the fix.

jmartin-sul commented on app/services/preserved_object_handler_results.rb:
```diff
@@ -0,0 +1,112 @@
+  end
+
+  # used when updates wrapped in transaction fail, and there is a need to ensure there is no db updated result
+  def remove_db_updated_result
```
nitpicking, but i would prefer pluralizing the name since it's possible multiple results will get removed.

ndushay replied:
agree - will fix.

jmartin-sul commented on app/services/preserved_object_handler.rb:
```diff
@@ -277,124 +232,66 @@ def update_pc_validation_timestamps(pres_copy)
+      handler_results.add_result(PreservedObjectHandlerResults::ARG_VERSION_LESS_THAN_DB_OBJECT, db_object.class.name)
     elsif incoming_version > db_object.send(version_symbol)
-      results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, db_object.class.name)
+      handler_results.add_result(PreservedObjectHandlerResults::ARG_VERSION_LESS_THAN_DB_OBJECT, db_object.class.name)
```
was changing the result code here intentional?  seems like this should stay as it was?  but also this didn't seem to break the tests, so i guess either i'm missing something or this is a test case we missed?

ndushay replied:
Actually, it's a test case I fudged, so I'm delighted you caught it.

https://github.com/sul-dlss/preservation_catalog/blob/7120f7411f70ef97f9d479373475d614340a85d5/spec/services/preserved_object_handler_update_version_spec.rb#L169-L179

ndushay commented:
@jmartin-sul I think I addressed your concerns with 3 follow-on commits;  I have a slight preference to squash commits if you are good with the resulting code.

jmartin-sul commented:
thanks, @ndushay!  all looks fine to me now.
